src/Elastic.Apm/Report/PayloadSenderV2.cs

```diff
@@ -0,0 +1,250 @@
+		}
+
+		private bool _isProcessingLoopRunning = true;
+		private Task _worker;
+
+		public void QueueTransaction(ITransaction transaction)
+		{
+			_eventQueue.Post(transaction);
+			_eventQueue.TriggerBatch();
+		}
+
+		public void QueueSpan(ISpan span) => _eventQueue.Post(span);
+
+		public void QueueError(IError error) => _eventQueue.Post(error);
+
+		/// <summary>
+		/// Flushes all the events and ends the loop that processes and sends the events.
+		/// This can be called only once and after that the instance won't process anything.
+		/// </summary>
+		internal async Task FlushAndFinishAsync()
+		{
+			_logger.LogDebug("FlushAndFinish called - PayloadSenderV2 will became invalid");
+			await _creation;
+			_isProcessingLoopRunning = false;
+			_eventQueue.TriggerBatch();
+
+			try
+			{
+				//Either this wins, or the other one from DoWork.
+				var queueItems = _eventQueue.Receive(timeout: TimeSpan.FromSeconds(1));
+				await ProcessQueueItems(queueItems);
+			}
+			catch
+			{
+				_logger.LogDebug("Reading from BatchBlock did not work, BatchBlock is empty.");
+			}
+
+			_batchBlockReceiveAsyncCts.Cancel();
+
+			try
+			{
+				await _worker;
+			}
+			catch (TaskCanceledException)
+			{
+				_logger.LogDebug("worker task cancelled");
+			}
+
+			_eventQueue.Complete();
+		}
+
+		private async Task DoWork()
+		{
+			_batchBlockReceiveAsyncCts = new CancellationTokenSource();
+			while (_isProcessingLoopRunning)
+			{
+				var queueItems = await _eventQueue.ReceiveAsync(_batchBlockReceiveAsyncCts.Token);
+				await ProcessQueueItems(queueItems);
+			}
+		}
+
+		private async Task ProcessQueueItems(object[] queueItems)
+		{
+			try
+			{
+				var metadata = new Metadata { Service = _service };
+				var metadataJson = JsonConvert.SerializeObject(metadata, _settings);
+				var json = new StringBuilder();
+				json.Append("{\"metadata\": " + metadataJson + "}" + "\n");
+
+				foreach (var item in queueItems)
+				{
+					var serialized = JsonConvert.SerializeObject(item, _settings);
+					switch (item)
+					{
+						case Transaction _:
+							json.AppendLine("{\"transaction\": " + serialized + "}");
+							break;
+						case Span _:
+							json.AppendLine("{\"span\": " + serialized + "}");
+							break;
+						case Error _:
+							json.AppendLine("{\"error\": " + serialized + "}");
+							break;
+					}
+				}
+
+				var content = new StringContent(json.ToString(), Encoding.UTF8, "application/x-ndjson");
+
+				var result = await _httpClient.PostAsync(Consts.IntakeV2Events, content);
+
+				if (result != null && !result.IsSuccessStatusCode)
+				{
+					var str = await result.Content.ReadAsStringAsync();
+					_logger.LogError($"Failed sending event. {str}");
+				}
+				if (result != null && result.IsSuccessStatusCode)
+				{
+					var sb = new StringBuilder();
+					sb.AppendLine("Sent items to server:");
+					foreach (var item in queueItems) sb.AppendLine(item.ToString());
+					_logger.LogDebug(sb.ToString());
```

Comment:
It might be worth checking if debug log level is enabled or better yet to add .LogXYZ overload that accepts lambda and executes only when level XYZ is enabled.

Reply:
Yeah, this is a general issue, most of our logging does too much when the log level is let's say `Error`. That should be a next iteration on logging... we definitely need those overloads, or some other solution. I'd keep it now as it is and deal with it later.